File: leetcode/go/962.maximum-width-ramp.go
func maxWidthRamp(nums []int) int {
    stack := make([]int, 0, len(nums))
    idx := 0
    for i, n := range nums {
        if len(stack) == 0 || nums[idx] >= n {
            idx = i
            stack = append(stack, i)
        }
    }
    ans := 0
    l := len(stack)-1
    for i:=len(nums)-1; i > -1; i-- {
        for l > -1 && nums[stack[l]] <= nums[i] {
            ans = max(ans, i - stack[l])
            l--
        }
    }
    return ans
}

File: leetcode/rust-leetcode/src/962.maximum-width-ramp.rs
impl Solution {
    pub fn max_width_ramp(nums: Vec<i32>) -> i32 {
        let mut stack = vec![];
        let mut idx = 0;
        for (i, &n) in nums.iter().enumerate() {
            if stack.is_empty() || nums[idx] >= n {
                stack.push(i);
                idx = i;
            }
        }
        idx = *stack.last().unwrap();
        let mut ans = 0;
        for (i, &n) in nums.iter().enumerate().rev() {
            while !stack.is_empty() && nums[idx] <= n {
                ans = ans.max(i - stack.pop().unwrap());
                if !stack.is_empty() {
                    idx = *stack.last().unwrap();
                }
            }
        }
        ans as _
    }
}

